src/Windows/Host/Client/Impl/BrokerServices/LocalStaticFileServer.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.IO;
+using Microsoft.Common.Core.Logging;
+
+namespace Microsoft.R.Host.Client.BrokerServices {
+    public class LocalStaticFileServer : StaticFileServerBase {
+        private static LocalStaticFileServer _server;
+        private static object _serverLock = new object();
+
+        public LocalStaticFileServer(IFileSystem fs, IActionLog log, IConsole console) : base(fs, log, console) {
+        }
+
+        public async Task<string> HandleUrlAsync(string urlStr, CancellationToken ct = default(CancellationToken)) {
+            await InitializeAsync(ct);
+
+            Log.WriteLine(LogVerbosity.Minimal, MessageCategory.General, Resources.Info_StaticFileServerStarted.FormatInvariant(Listener.Prefixes.FirstOrDefault()));
+            Console.WriteLine(Resources.Info_StaticFileServerStarted.FormatInvariant(Listener.Prefixes.FirstOrDefault()));
+
+            UriBuilder ub = new UriBuilder(urlStr);
+            return GetFileServerPath(ub.Path);
+        }
+
+        public static Task<string> CreateAsync(string url, IFileSystem fs, IActionLog log, IConsole console, CancellationToken ct = default(CancellationToken)) {
+            lock (_serverLock) {
+                _server = _server ?? new LocalStaticFileServer(fs, log, console);
+            }
+
+            return _server.HandleUrlAsync(url);
+        }
+
+        public override async Task HandleRequestAsync(HttpListenerContext context, CancellationToken ct) {
+            var uri = context.Request.Url.PathAndQuery;
+            if (!FileSystem.FileExists(uri) && uri.StartsWith("/")) {
```

Comment:
Ah, so it comes from the request. But doesn't this imply that there's basically no way to distinguish between relative and absolute file access in the protocol as used here? It feels like it should either always be stripping it, or never stripping it, but not trying to check if file exists or not first. Otherwise you get this weird "file at this absolute path if it exists, otherwise file at this relative path" semantics that can turn error conditions into invalid data being returned quietly.